dnsdist: Add an option to use several source ports toward a backend

rgacogne opened the pull request:
### Short description
<!-- Write a small description of what this Pull Request fixes or provides, including the issue #s -->
This is very useful if the backend is distributing queries based only on (source IP, source port, destination IP, destination port), which is for example the case of PowerDNS Recursor with several
threads, `reuseport` set and `pdns-distribute-queries` not set.

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [ ] compiled and tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

zeha commented on pdns/dnsdist-lua.cc:
```diff
@@ -216,6 +217,10 @@ void setupLuaConfig(bool client)
 			}
 
+			if(vars.count("sockets")) {
+			  numberOfSockets=std::stoi(boost::get<string>(vars["sockets"]));
```
maybe add a check > 0

rgacogne commented:
Added a check that the number of sockets is greater than 0, handle more gracefully a failed (re-)connection, and updated the file descriptors check to account for more than one socket per backend.

rgacogne commented:
Note that the current implementation will not perform well with a large number of sockets per backend since it currently uses `poll()`. It would be better to use our existing multiplexer classes, but it requires:
- removing / adding the socket descriptors to the multiplexer on reconnection ;
- refactoring the existing code to support the multiplexer callback interface, or adding a new function to the multiplexer interface that would return a list of 'ready' file descriptors.

rgacogne commented:
Pushed an additional commit using our existing multiplexer code. I went with the second option described above, adding a new function to the multiplexer interface that returns a list of 'ready' file descriptors. This code needs a serious review before merging!
